### ISSUE TITLE
Mypy ignores typing.no_type_check for nested function redefintions.

### ISSUE THREAD

**Issue by bharel** [python/mypy#10002]
**Bug Report**

Mypy ignores no_type_check, specifically for nested function redefinitions.

![image](https://user-images.githubusercontent.com/742359/106412961-929fb200-6451-11eb-8e0d-61245451ddde.png)

**To Reproduce**

1. Create a nested function.
2. Create another nested function in the same name right after it.
3. Attempt to suppress it with `@typing.no_type_check`.

I've tried to deal with it in many ways, including opening a [different](https://github.com/python/mypy/issues/10001) bug, to no avail.

**Your Environment**

- Mypy version used: 0.800
- Mypy command-line flags:  --disallow-untyped-defs, --show-column-numbers, --follow-imports=silent, --ignore-missing-imports, --check-untyped-defs
- Mypy configuration options from `mypy.ini` (and other config files): Nope
- Python version used: 3.9.0
- Operating system and version: Windows 10

**Comment by terencehonles** [comment]
I would expect the decorator applies to the body not to the name itself. For that I would expect you would still need `# type: ignore[no-redef]` but it looks like you also have the decorator on the enclosing function so I would expect you not to need the type comment.

For kicks do you want to test if the name `test` is the only issue or does it not descend into enclosed functions? Actually looking at https://github.com/python/mypy/issues/10001 you are suggesting it is type checking the enclosed function.

I would expect you would need

```python

@no_type_check
def wrapper(...):
    # lines don't need type: ignore

    @no_type_check
    def wrapped(...):  # type: ignore[no-redef]
        # lines don't need type: ignore
```

If you indeed need the `# type: ignore[no-redef]` on the enclosed function I would suggest that's a bug since it should be covered by the `@no_type_check` decorator on `wrapper`, but I'm not sure `mypy` should default to not type checking recursively. If you are incrementally adding types you may add typing to the enclosed function and it might be better to be explicit and add the `@no_type_check` decorator you were attempting to add.

**Comment by bharel** [comment]
Recursive or not is something we can discuss over, but the redefinition
linting is a bug for sure.

**Comment by JelleZijlstra** [comment]
I would agree that `@no_type_check` should suppress all errors within the body of a function, including in nested functions and redefinitions. I haven't checked but it might be a fairly easy fix.
